### ISSUE TITLE
RFC: Add ComponentStore for managing local component state

### ISSUE THREAD

**Issue by spierala** [spierala/mini-rx-store#59]
A Component Store could be beneficial to manage state at a local (component) level.

But let's see what we have already in MiniRx...

MiniRx Store and Feature Store manage state at a global level:

**MiniRx Store**
- Is a classic Redux Store which manages a global state object
- In most cases the global state has the lifespan of the application
- Redux with its Actions and Reducers allows us to use Redux Dev Tools
- Store Extensions like the UndoExtension or ImmutableExtension are MetaReducers, which require Redux too

Read more in the MiniRx Store docs here: https://mini-rx.io/docs/redux
  
**MiniRx Feature Stores** 
- Feature Store state is also registered in the global state object
- Feature Stores use Redux under the hood, so we can also use Redux Dev Tools and benefit from other Store Extensions (e.g. UndoExtension, ImmutableExtension)
- Feature Stores allow to interact directly with a specific feature state with a more easy API (no Redux Boilerplate :))
- API: select, setState, undo, effect, get state()
- Feature Stores are destroyable, therefore they can take the role of a Component Store

Read more in the MiniRx Feature Store docs here: https://mini-rx.io/docs/fs-quick-start

### Feature Stores as Component Store
MiniRx Feature Stores come close to a Component Store since they are destroyable.
See an example in the Angular Demo here: https://angular-demo.mini-rx.io/#/counter

The "counter" Feature Store looks like this:
https://github.com/spierala/mini-rx-angular-demo/blob/main/src/app/modules/counter/state/counter-state.service.ts

Note that it has to provide a unique feature key:
```
constructor() {
    super('counter-' + id++, initialState);
}
```

The feature key is used to register the counter states into the global state object:
![Screenshot 2021-11-09 at 20 40 24](https://user-images.githubusercontent.com/1272446/140993282-fc42d1cf-2272-4092-8a4c-5e72edf48231.png)

The counter component provides its own instance of the counter-state.service. See here: 
https://github.com/spierala/mini-rx-angular-demo/blob/main/src/app/modules/counter/counter/counter.component.ts#L9

When the counter components are destroyed then the corresponding feature states are removed from the global state object.

### Feature Stores as Component Store PROs & CONs

**PROs**
- Existing functionality can be reused (the Feature Store class) to have a store with the lifespan of the component
- The component state can be inspected with Redux Dev Tools, we can undo state changes, etc
- Feature Stores automatically use the extensions which are provided for the global store (e.g. ImmutableExtension)
- Since the component state is registered in the global state we can easily write selectors which combine other global state with the "component state"

**CONs**
- It is necessary to provide a unique feature key when instantiating the Feature Store
- When we register / unregister hundreds of Feature Stores at the same time there is a visible performance impact (only at the moment when the register/unregister happens, afterwards it is fine)
- If there are a lot of Feature Stores used as Component Stores they might "pollute" the global state object with a lot of state which is meant as local state. 
- Local component state changes trigger emissions on the global state object which is odd and less performant.

### MiniRx Component Store?

What could it look like?

- Same API as Feature Stores (select, setState, effect, undo, get state())
- No need to provide a unique feature key

**Mini Redux?**
- The Undo functionality would require that the Component Store is a little Redux Store itself
- If the Component Store is a little Mini Redux we could easily add support for some of the existing MiniRx extensions: ImmutableExtension, UndoExtension, LoggingExtension. Not sure about the Redux Dev Tools extension.

### MiniRx Component Store PROs & CONs

**PROs**
- Component Store would be more performant than Feature Store (mostly if you have hundreds of components using Component Store)

**CONs**
- It could be confusing for the user of MiniRx to see another Store variant which is very similar to FeatureStore ("when should I use which one?"). We would have 3 variants: Store, Feature Store, Component Store
- There will be a lot of duplication in the docs since most of the API is identical with FeatureStore
- MiniRx would loose its focus on global state-management ( thats maybe a marketing thing ;) but I think a young library like MiniRx should have focus to become successful )
- Thinking about the future: enitityAdapter for Redux and a EntityFeatureStore? Would we need then also a EntityComponentStore? 

### Make the best of Feature Store?
**Internal unique id creation?**
We could try to make the registration of multiple Feature Stores more straightforward, especially the unique id thing, which could be handled internally

**Detached Mode?**
Prevent Feature Store from the registration on the global Redux store. Instead the Feature Store could fall back to a local Redux.
Again we need Redux mostly to keep the undo functionality (using the UndoExtension).

### Non Goals
- Be a copy of ngrx/component-store

**Comment by nartc** [comment]
Thanks for opening this RFC.

TLDR; Dedicated Component Store is nice to have but not necessary. `FeatureStore` is probably a better solution with some rework

Initially, I think a dedicated Component Store solution would be nice to have for the following reasons:
- Easy to get started with (just a glorified Service-with-a-Subject)
- Pathway to migrate to `FeatureStore` if needs Global functionality

But the CONS you've outlined in the RFC are all valid. The overlap between Component Store and Feature Store will be large and will cause a lot of confusion for the consumers which, to me, I think is even worse than NOT having a Component Store at all.

That said, I'd still like to see a local-focused solution in terms of State Management from MiniRx. Maybe Feature Store can be reworked to be more local-focused? Several concerns come to mind:
- Selector usage needs to be simplified for Local State management. Memoized selectors are great but might be overkill for a Local State solution.
- Ease of instantiating multiple instances of the same Store.
- Utilize the existing Extension?
- (Will be really nice) if Redux Extension can be worked in, this is currently missing from other Local State solutions (RxState, ComponentStore)

PS: Should this be a Discussion rather than an Issue?
